### ISSUE TITLE
download-frozen-image-v2.sh (branch 19.03/master) does not download from non-Docker registries

### ISSUE THREAD

**Issue by afflom** [moby/moby#40857]
**Description**
When trying to download container images from non-docker registries (k8s.gcr.io, quay.io, registry.centos.org), download-froxen-image-v2.sh returns "curl: (22) The requested URL returned error: 401 Unauthorized"

**Steps to reproduce the issue:**
1. ```wget https://github.com/moby/moby/blob/19.03/contrib/download-frozen-image-v2.sh```
2. ```chmod +x download-frozen-image-v2.sh```
3. ```./download-frozen-image-v2.sh "kube-apiserver" k8s.gcr.io/kube-apiserver:v1.18.2```

**Describe the results you received:**
```curl: (22) The requested URL returned error: 401 Unauthorized```

**Describe the results you expected:**
I expected successful completion of the script with the output looking similar to:
```
Downloading 'k8s.gcr.io/kube-apiserver:v1.18.2' (1 layers)...

######################################################################## 100.0%

Download of images into 'kube-apiserver' complete.
Use something like the following to load the result into a Docker daemon:
  tar -cC 'kube-apiserver' . | docker load
```

**Additional information you deem important (e.g. issue happens only occasionally):**
I am able to successfully pull images with download-frozen-image-v2.sh from Docker Hub

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           19.03.8
 API version:       1.40
 Go version:        go1.12.17
 Git commit:        afacb8b
 Built:             Wed Mar 11 01:27:04 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.8
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.17
  Git commit:       afacb8b
  Built:            Wed Mar 11 01:25:42 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Output of `docker info`:**

```
lient:
 Debug Mode: false

Server:
 Containers: 17
  Running: 17
  Paused: 0
  Stopped: 0
 Images: 9
 Server Version: 19.03.8
 Storage Driver: overlay2
  Backing Filesystem: <unknown>
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: systemd
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 3.10.0-957.27.2.el7.x86_64
 Operating System: CentOS Linux 7 (Core)
 OSType: linux
 Architecture: x86_64
 CPUs: 3
 Total Memory: 7.638GiB
 Name: 1588216178.test
 ID: ZOYK:SR43:OIMO:ZT36:WBGK:ALKP:WDHV:4NYF:QPN5:GLKF:ZFPB:TPFY
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Docker is not installed on the device that I am using download-frozen-image-v2.sh on.

**Comment by afflom** [comment]
I have an idea about how to make a command line flag to override these:
```
registryBase='https://registry-1.docker.io'
authBase='https://auth.docker.io'
authService='registry.docker.io'
```
But I'm not quite sure about how to get this info for other registries or if it is all required to pull public images. If someone could point me in the right direction, I can do the work on the script and submit the a PR for the change myself. Any help would be appreciated.

**Comment by thaJeztah** [comment]
Thanks for reporting; these scripts were created for use in our CI, so definitely have not been designed to be a "universal" solution.

That said, if only minimal changes are needed (and not making the script super-complex to address other uses), I think that would be ok.

Perhaps something like this;

```patch
diff --git a/contrib/download-frozen-image-v2.sh b/contrib/download-frozen-image-v2.sh
index 283a9ab34d..ef925b9008 100755
--- a/contrib/download-frozen-image-v2.sh
+++ b/contrib/download-frozen-image-v2.sh
@@ -43,9 +43,9 @@ if [ "$(go env GOHOSTOS)" = 'windows' ]; then
        fi
 fi

-registryBase='https://registry-1.docker.io'
-authBase='https://auth.docker.io'
-authService='registry.docker.io'
+registryBase="${registryBase:-'https://registry-1.docker.io'}"
+authBase="${authBase:-'https://auth.docker.io'}"
+authService="${registryBase:-'registry.docker.io'}"

 # https://github.com/moby/moby/issues/33700
 fetch_blob() {
```


Then the variables can be overridden before the script is called; something like

```bash
registryBase='https://registry-1.example.com' authBase=... authService=... ./download-frozen-image-v2.sh ...
```


@tianon WDYT?

**Comment by afflom** [comment]
Thanks for taking a look at this. Interestingly, many people reference this script as the best solution to pull container images with minimal dependencies. You guys may have accidentally created something that is very useful to the masses.

With your proposed change would I be able to omit the authBase and authService if pulling public images like this?:
```
registryBase='https://registry-1.example.com'  ./download-frozen-image-v2.sh ...

```

**Comment by thaJeztah** [comment]
> You guys may have accidentally created something that is very useful to the masses.

Ha! Yes, it's sometimes interesting to see where things end up. I can see the script definitely being useful yes.

> With your proposed change would I be able to omit the authBase and authService if pulling public images like this?:

My change was basically;

```bash
fooVar="${fooVar:-'default value'}"
```

So, if `fooVar` is set on the command line it would override the default value (and if it's not set on the command line, the default value would be used)

So, if the question is; would it skip authentication if it's not set; _no_.
Would that be possible? Perhaps, but I haven't dug deep into the script, perhaps a conditional check could be added (if `authBase` is empty, then skip the authentication step); at a quick glance, that would be slightly more involved (e.g. `fetch_blob()` expects a token to be passed, and adds it to the headers, so steps like that also need to be made conditional.

Feel free to open a pull request for review/discussion (it may be easier to discuss changes when looking at actual code that makes those changes) if you have suggestions for that. If possible, I would still like to keep the script "simple" (e.g., if overriding an environment variable does the job, don't add flags/arguments)

**Comment by afflom** [comment]
Thank you so much for your help on this. I'll get started on it and submit when I have something (hopefully) working and simple.

**Comment by tianon** [comment]
Yeah, I think auth is going to be the most difficult part of this -- each registry handles auth in a slightly different way (auth is intentionally not part of the OCI distribution spec, for example), so I think the best we could do here is make auth either swappable for registries that use it the same way Docker Hub does or optional for registries that don't require auth (like `mcr.microsoft.com` for example).

**Comment by afflom** [comment]
I'm testing the cURLs with different authentication and trying to understand the required parts for unauthenticated pulls. I was then going to put some simple logic in where the if there is a mismatch between the registry and the default (docker) auth provider, the image pull would be sent as an unauthenticated request.

**Comment by matya** [comment]
Hi,
I have hit the same problem with using private registries. Fortunately there was not too much magic involved, so I have forked the script and added some features, like automatically determining where it needs to pull the auth token from.
https://github.com/matya/download-docker-image/blob/master/download-docker-image.sh
Feedback is welcome.

**Comment by thaJeztah** [comment]
@matya as mentioned above (haven't looked at the specific changes in your version); if the changes are reasonable, feel free to open a pull request to upstream your modifications! 🤗

**Comment by matya** [comment]
I have more or less rewritten major parts of it and also removed some code that required the dependency to go, so unless we can find some non-go replacement for it, it will probably loose features you already have and require...

**Comment by thaJeztah** [comment]
No worries; feel free to push a PR if you want at some point; pull requests are "cheap" (we can always close them if it's not a good fit; no harm done in that case 👍)

**Comment by matya** [comment]
As far as the base-usage is considered, I have tested the outcome of downloading the [your use-case](https://github.com/moby/moby/blob/e184eac73a87a55bd77caf988ea9dea4b209e471/Dockerfile#L91) with the new script and it produces the same files down to the bit, so that would still be handled properly and could be used as a drop-in replacement with some changes regarding parametrisation...

I am completely new to the moby project and as far I have seen this script is in use in some build process, but also has some code to handle Windows+Bash4 combo and is being referenced from some go files by name, which is currently out of the scope I can edit and test.

My capacities for testing it regarding free time and also proper build environment are also quite limited outside of work... However, I feel the feature set if something lightyears away what is in the current version (like support for any external registry or proper authentication etc.), and those few who actually require such an uncommon task to download without docker pull would benefit from it and make their life easier, so I will try to go down that road on the weekend...

**Comment by afflom** [comment]
@matya you beat me to it, but I'm glad. Can you please open a concurrent PR and reference this issue? Like you said, many people will benefit from your work. Thank you.

**Comment by tianon** [comment]
What if we instead keep this script specific to the Moby use case and updated it to link to your more advanced version for users who want to use it to do more than what we need here?

**Comment by matya** [comment]
Fine with me

**Comment by tianon** [comment]
Awesome!  I'm thinking something close to the top (probably in both the `v1` and `v2` script to increase the visibility?) that points to https://github.com/matya/download-docker-image as the suggested endpoint for users who need something more than what we need here:

```bash
#
# This script was developed for use in Docker's CI, and as such the use cases and usability are (intentionally) limited.
#
# For users who have a similar need but require more flexibility/functionality, Janos Mattyasovszky maintains a fork of this script at:
#
#   https://github.com/matya/download-docker-image
#
```

@thaJeztah what are your thoughts?  Does that seem reasonable?

**Comment by afflom** [comment]
Just a question: why not make this a v3?

**Comment by tianon** [comment]
The `v1` and `v2` there refer to the version of the registry API they interact with, not the version of the script. :sweat_smile:

**Comment by berlinsaint** [comment]
A question, how to use docker registry mirror by this script?

**Comment by afflom** [comment]
@berlinsaint If not pulling from docker hub, I think the recommendation is this: https://github.com/matya/download-docker-image

**Comment by berlinsaint** [comment]
Nice Job ,  I found some feature that i want to modify.  And i finally find that every registry may implete differently and will cause to many error such as 404 , if not using docker hub

**Comment by matya** [comment]
I can also recommend [Scopeo](https://github.com/containers/skopeo) that I found only after writing my version and have since then switched over to this because of the even more compactness of it.
